### PR TITLE
[GPU] Do not use usm_host when network output tensor size is large

### PR DESCRIPTION
**Pull request by isanghao**
### Details:
 - For dGPU including A770, when network output size is large, performance is better with explicit copy then writing data to usm_host.
 - Allow usm_host access for input
 - Next-gen dGPU, write to usm_device and copy it to usm_host
 - For DG2, write to usm_device only for large output

### Tickets:
 - 157439

### CODE REVIEW

File: src/plugins/intel_gpu/src/plugin/sync_infer_request.cpp [hunk 1]
```diff
@@ -790,7 +799,7 @@ std::vector<cldnn::event::ptr> SyncInferRequest::prepare_input(const std::string
         } else {
             m_plugin_inputs[input_idx] = user_tensor_wrapper;
         }
-    } else if (is_usm_host_tensor && !convert_needed && can_use_usm_host(engine)) {
```
[the remark this answers is not in the document]
Comment (isanghao) [reply src/plugins/intel_gpu/src/plugin/sync_infer_request.cpp:793]: yes, it is intentional and it was actually helpful to improve the performance though I did not test on PVC.

File: src/plugins/intel_gpu/src/plugin/sync_infer_request.cpp [hunk 2]
```diff
@@ -32,12 +32,17 @@
...
+    GPU_DEBUG_IF(debug_config->use_usm_host == 1) { return true; }
+    GPU_DEBUG_IF(debug_config->use_usm_host == 2) { return false; }
+
     auto can_use_usm = engine.use_unified_shared_memory();
 
     const auto& device_info = engine.get_device_info();
     if ((device_info.gfx_ver.major == 12 && device_info.gfx_ver.minor == 60) ||
-        (device_info.gfx_ver.major >= 20 && device_info.dev_type == cldnn::device_type::discrete_gpu)) {
+        (device_info.gfx_ver.major >= 20 && device_info.dev_type == cldnn::device_type::discrete_gpu) ||
+        (device_info.dev_type == cldnn::device_type::discrete_gpu && total_output_bytes > 4*1048576)) {
```
[the remark this answers is not in the document]
Comment (isanghao) [reply src/plugins/intel_gpu/src/plugin/sync_infer_request.cpp]: applied, thanks!

File: src/plugins/intel_gpu/include/intel_gpu/plugin/sync_infer_request.hpp [hunk 3]
```diff
@@ -118,6 +118,7 @@ class SyncInferRequest : public ov::ISyncInferRequest {
 
     void init_mappings();
     bool is_batched_input(const ov::Output<const ov::Node>& port) const;
+    uint64_t total_output_bytes = 0;
```
[the remark this answers is not in the document]
Comment (isanghao) [reply src/plugins/intel_gpu/include/intel_gpu/plugin/sync_infer_request.hpp:121]: 1. Yes, actually there is room for further fine-tuning this approach, including BMG behavior. This is a quick&simple heuristic. Can I put that as a further work in backlog for the future work?
2. For behavior, you are right. In dynamic-shaped model, the output size will be calculated to 0. But more importantly, output buffer will be created separately and `can_use_usm_host` does not affect the execution flow. Ireq's output buffer will be allocated during `wait()` and explicit copy will happen.

[the remark this answers is not in the document]
Comment (sshlyapn) [reply src/plugins/intel_gpu/include/intel_gpu/plugin/sync_infer_request.hpp:121]: Sure, it's just a suggestions for further improvements

Oh, yeah, as for dynamic outputs, this change doesn't affect them directly, sorry for the false alarm!
I briefly checked dynamic outputs again, and it seems there is still room for further improvements. As you mentioned, we always apply explicit memory copy for dynamic outputs, but internally, in cldnn::network, we allocate lockable memory for the last node (`is_output_buffer` check):
https://github.com/openvinotoolkit/openvino/blob/9d36703ab9f695b27c01c68b1f3464de7ba3026f/src/plugins/intel_gpu/src/graph/primitive_inst.cpp#L2366-L2367

And if this is happening for all the dynamic outputs, then we might be facing the same issue you're addressing (when last kernel uses usm_host as dst memory). So, maybe we need to take a closer look at this, and if my observations are correct, address this case too

**Review by sshlyapn — APPROVED** [review]

**Pull request merged**